bake-kdf/src/lib.rs

```diff
@@ -0,0 +1,167 @@
+        6 => Ok([
+            k[0],
+            k[1],
+            k[2],
+            k[3],
+            k[4],
+            k[5],
+            k[0] ^ k[1] ^ k[2],
+            k[3] ^ k[4] ^ k[5],
+        ]),
+        // if n = 8, then K = K_1 || K_2 || K_3 || K_4 || K_5 || K_6 || K_7 || K_8
+        8 => Ok([k[0], k[1], k[2], k[3], k[4], k[5], k[6], k[7]]),
+        // otherwise, return error
+        _ => Err(InvalidKeyLength),
+    }
+}
+
+/// Key repetition algorithm described in STB 34.101.34-2020 8.1.3
+/// Input data:
+/// Expanded key: K = K_1 || K_2 || ... || K_n,
+/// where K_i is 32-bit word
+/// n = 4, 6, 8
+/// Data: D = D_1 || D_2 || D_3,
+/// where D_i is 32-bit word
+/// Initialization vector: I = I_1 || I_2 || I_3 || I_4,
+/// where I_i is 32-bit word
+/// Output data:
+/// Repetition key: Y = Y_1 || Y_2 || ... || Y_n,
+/// where Y_i is 32-bit word
+pub fn belt_keyrep<const M: usize>(
+    x: &[u32],
+    d: &[u32],
+    i: &[u32],
+    out: &mut [u32],
+) -> Result<(), Error> {
+    let n = x.len() * 32;
+    if (n != 128) && (n != 192) && (n != 256) && (M != 128) && (M != 192) && (M != 256) {
+        return Err(InvalidKeyLength);
+    }
+    if d.len() != 3 || i.len() != 4 {
+        return Err(InvalidDataSize);
+    }
+
+    let r: u32 = match (n, M) {
+        (128, 128) => 0xC8BA94B1,
+        (192, 128) => 0x12D6E35B,
+        (192, 192) => 0xFFC0B05C,
+        (256, 128) => 0x1ADC2BE1,
+        (256, 192) => 0x3876ABC1,
+        (256, 256) => 0x7B653CF3,
+        _ => unreachable!(),
+    };
+
+    let s = belt_keyexpand(x).map_err(|_| InvalidKeyLength)?;
+    let mut d = [d[0], d[1], d[2]];
+    let mut i = [i[0], i[1], i[2], i[3]];
+
+    d.iter_mut().for_each(|x| *x = u32::to_be(*x));
+    i.iter_mut().for_each(|x| *x = u32::to_be(*x));
+
+    let (_, s) = belt_compress([r, d[0], d[1], d[2]], i, s);
+
+    match M {
+        128 => {
+            let mut y = [s[0], s[1], s[2], s[3]];
+            y.iter_mut().for_each(|x| *x = u32::to_be(*x));
+            out.copy_from_slice(&y);
+        }
+        192 => {
+            let mut y = [s[0], s[1], s[2], s[3], s[4], s[5]];
+            y.iter_mut().for_each(|x| *x = u32::to_be(*x));
+            out.copy_from_slice(&y);
+        }
+        256 => {
+            let mut y = [s[0], s[1], s[2], s[3], s[4], s[5], s[6], s[7]];
+            y.iter_mut().for_each(|x| *x = u32::to_be(*x));
+            out.copy_from_slice(&y);
+        }
+        _ => unreachable!(),
+    }
+
+    Ok(())
+}
+
+/// bake-kdf algorithm described in STB 34.101.66-2014 8.1.4
+pub fn bake_kdf(x: &[u8], s: &[u8], c: u128) -> Result<[u32; 8], Error> {
+    let mut hasher = BeltHash::default();
+    hasher.update(x);
+    hasher.update(s);
+    let y = hasher.finalize_fixed();
+
+    let d: [u32; 3] = [0xFFFFFFFF, 0xFFFFFFFF, 0xFFFFFFFF];
+
+    let mut c = to_u32::<4>(&c.to_be_bytes());
+    c.reverse();
```

Comment:
Are you sure the reversal is needed on BE targets? Ideally, we should test correctness on BE targets using either MIRI or cross.

Reply:
I will check tomorrow, thank you. Not sure in it

Reply:
@newpavlov fiex.